### PR TITLE
SPIR-V shader support

### PR DESCRIPTION
**Pull request by zakorgy**
Modified the shaders for SPIR-V compilation.
Extended the build script:
1. Shader concatenation 
2. Compile shaders to SPIR-V (glslangValidator) binary and validate it (spirv-val).

### CODE REVIEW

**Comment by kvark** [comment]
What we really need here is to minimize the changes to original shaders. Can we automatically map I/O of the shaders, for example? We need some automated processing of the original into SPIRV.

**Comment by dati91** [comment]
Yeah, for sure.

Our long-term plan is to generate the binding location from `build.rs`, save these (like `shaders.rs` map), and in the code we can generate descriptors from that.

Our short-term plan is to get to a working version (with this pr) so we can continue the work on the renderer/device side.

We should create an issue for tracking the progress for this.

File: webrender/build.rs [hunk 1]
```diff
@@ -2,13 +2,327 @@
...
+#[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+const SHADER_VERSION: &'static str = "#version 300 es\n";
+#[cfg(feature = "vulkan")]
+const SHADER_VERSION_VK: &'static str = "#version 450\n";
+#[cfg(feature = "vulkan")]
+const VK_EXTENSIONS: &'static str = "#extension GL_ARB_shading_language_420pack : enable\n\
+                                     #extension GL_ARB_explicit_attrib_location : enable\n\
+                                     #extension GL_ARB_separate_shader_objects : enable\n";
+
+const CACHE_FEATURES: &[&str] = &[""];
```
Comment (dati91) [inline webrender/build.rs]: This looks needless, or maybe call it empty and use it when you use `&[""]`

File: webrender/build.rs [hunk 2]
```diff
@@ -2,13 +2,327 @@
...
+const CLIP_FEATURES: &[&str] = &["TRANSFORM"];
```
Comment (dati91) [inline webrender/build.rs]: Why do you have these feature `const`s if the others are inlined?

File: webrender/build.rs [hunk 3]
```diff
@@ -2,13 +2,327 @@
...
+            "YUV_REC709,TRANSFORM",
+            "INTERLEAVED_Y_CB_CR,YUV_REC709,TRANSFORM",
+        ],
+    },
+    Shader {
+        name: "ps_text_run",
+        source_name: "ps_text_run",
+        features: &["", "TRANSFORM", "GLYPH_TRANSFORM"],
+    },
+    Shader {
```
Comment (dati91) [inline webrender/build.rs]: Put a comment here (e.g. `// Debug shaders`)

File: webrender/build.rs [hunk 4]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
         // if someone is building on a network share, I'm sorry.
         let full_name = full_name.replace("\\\\?\\", "");
         let full_name = full_name.replace("\\", "/");
+        shader_map.insert(shader_name.clone(), full_name.clone());
         write!(
             shader_file,
             "    h.insert(\"{}\", include_str!(\"{}\"));\n",
-            shader_name,
-            full_name
```
Comment (dati91) [inline webrender/build.rs]: this can stay the way it was

File: webrender/build.rs [hunk 5]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+    let write_json = file_name.ends_with(".vert");
+    let mut attribute_descriptors = Vec::new();
+    let mut descriptor_set_layouts = Vec::new();
+    let mut vertex_offset = 0;
+    let mut instance_offset = 0;
+    let mut pipeline_requirmenets = HashMap::new();
+    let is_blur_shader = file_name.starts_with("cs_blur");
+    let is_clip_shader = file_name.starts_with("cs_clip");
+
+    // Mapping from glsl sampler varaible name to a tuple,
```
Comment (dati91) [inline webrender/build.rs]: `varaible` -> `variable`

File: webrender/build.rs [hunk 6]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+    // in wich the first item is the corresponding expression used in vulkan glsl files,
```
Comment (dati91) [inline webrender/build.rs]: `wich` -> `which`

File: webrender/build.rs [hunk 7]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+                            "ty": "Sampler",
+                            "count": 1,
+                            "stage_flags": "All",
+                        }));
+                    }
+                    new_data.push_str(&layout_str);
+                    binding += 1;
+                }
+
+            // Replace non sampler uniforms with a structure.
```
Comment (dati91) [inline webrender/build.rs]: `non sampler` -> `non-sampler`

File: webrender/build.rs [hunk 8]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+            // This is placed to the position of the last non uniform variable (uDevicePixelRatio).
```
Comment (dati91) [inline webrender/build.rs]: `non uniform` -> `non-uniform`

File: webrender/build.rs [hunk 9]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+                     \t\tmat4 uTransform;       // Orthographic projection\n\
+                     \t\tuniform float uDevicePixelRatio;\n\
+                     \t\t// A generic uniform that shaders can optionally use to configure\n\
+                     \t\t// an operation mode for this batch.\n\
+                     \t\tuniform int uMode;\n\
+                     \t};\n",
+                );
+            }
+
+        // Adding location info for non uniform variables.
```
Comment (dati91) [inline webrender/build.rs]: `non uniform` -> `non-uniform`

File: webrender/build.rs [hunk 10]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+                                    "binding": 0,
+                                    "element": {
+                                        "format": format,
+                                        "offset": vertex_offset,
+                                    }
+                                }));
+                            vertex_offset += offset;
+                        }
+                        "aData0" | "aData1" => {
+                            // We have to leave these variables from blur and clip shaders
```
Comment (dati91) [inline webrender/build.rs]: Not sure what this comment means

File: webrender/build.rs [hunk 11]
```diff
@@ -27,16 +341,374 @@ fn write_shaders(glsl_files: Vec<PathBuf>, shader_file_path: &Path) {
...
+            .unwrap() != 0
+        {
+            println!("Error while validating spirv shader: {:?}", file_name);
+            process::exit(1)
+        }
+    }
+    json!(requirements)
+}
+
+fn write_to_json(requriements: serde_json::value::Value, out_dir: &str) {
```
Comment (dati91) [inline webrender/build.rs]: This name is a bit misleading, because you use the name `write_json` before.
Maybe `write_json_to_file`?

**Pull request merged**